### PR TITLE
Issue 134 implement tuners metamodels acquisition

### PR DESCRIPTION
**Pull request by pvk-developer**
- Implemented `GaussainProcessMetaModel`.
- Implemented `AcquisitionFunctions` of : `ExpectedImprovementFunction` and `NumpyArgSort` (used for maximize and minimize).
- Implemented `GPTuner` (based on `NumpyArgSort` and `GPEiTuner` (based on `ExpectedImprovement`).
- Update to the project documentation.
- Added the old `selection` modules.
- Unit tests.

### CODE REVIEW

**Comment by csala** [comment]
Thanks for the feedback @micahjsmith

Some general comments:
 
> * We should try to improve the organization of modules within `btb.selection` to match the
>   structure of `btb.tuning`. I understand that this was not within the scope of the
>   current re-design or PR. However, we should take this opportunity to fix that. For
>   example, `btb.selection.uniform` -> `btb.selection.selectors.UniformSelector`.

I disagree with this. As you say, the selectors are out of the scope of this PR. They were just mistakenly removed several commits ago and now they have been re-introduced exactly as they were before, but without any modifications. So I see no advantage in doing the selector changes here, compared to doing it on a separated PR right after.

For this reason, and since this is a big enough PR already, can we stick to the original plan and focus only on the tuning side to have this in asap?

> * The choice of which objects are re-exported at a higher level seems somewhat arbitrary, as
>   I comment in-line in one place. I propose: any object that we expect an end-user (user
>   of tuning, not developer of tuners) to use in an application should be re-exported at
>   the top-level `btb` module. Thus we should enable `from btb import UniformTuner, UniformSelector, IntHyperParam`.

I agree with the part of not making arbitrary choices: let's have all the user-level elements at the same level.
However, I would still make a clear distinction between tuning and selection families.
Just like you have to do `from numpy.random import randint, random` instead of `from numpy import randint, random`, or `from sklearn.ensemble import RandomForestClassifier` instead of `from sklearn import RandomForestClassifier`.

So, can we make it be `from btb.tuning import IntHyperParam, GPTuner, Tunable...` and `from btb.selection import ...`?

> * I see the pattern `import logging; logger = logging.getLogger('btb')` repeatedly, why not
>   instead create a `btb.util.log` module from which an instance of this logger can be
>   imported? This ensures consistent logging standards across the project. A reason to
>   _not_ use this paradigm would be if each module used logging differently (i.e. used
>   different loggers), but that is not the case (i.e. each module uses the `'btb'` logger).

This is in the `selection` module, so let's ignore this for now.

But, in any case, when we do changes there, I totally agree that importing a fixed name logger is wrong, but I see no difference between this and importing a fixed name logger from a utils module.

The way to go is always using the standard `LOGGER = logging.getLogger(__name__)`, which is the only option that gives complete control to the user to select which level each module has, as well as information of what is logging what directly at the logger name level.

> * the full name `AcquisitionFunction` seems to be avoided in names of classes, probably
>   because it is too long, so we should use `Acquisition` consistently, i.e.
>   `BaseAcquisition` rather than `BaseAcquisitionFunction`.

:+1: 

> * there are a variety of typos and grammatical errors in docstrings which is perfectly fine
>   :) perhaps look into whether you can configure your editor to do spellcheck/grammarcheck
>   in docstrings. I will go through in a separate commit after this process to clean things
>   up.

:+1: 

> * feedback on notebooks
>   * instead of including call to `os.system('pip install matplotlib')`,
>     include a simple requirements.txt within the notebooks directory that contains
>     notebooks-specific requirements: matplotlib, tqdm, etc.. Alternately (and Carles will probably love this
>     solution), create an `extras_require` entry in setup.py with key=`examples` or
>     `notebooks`

Indeed, I like the later. @pvk-developer let's call it `examples`, and let's add both `matplotlib` and `jupyter` there, as well as any other thing that is only used in the examples.

>   * re-run the notebooks, as I see that `scale` is still a kwarg passed to GPTuner which
>     does not appear in the source

:+1:

**Comment by micahjsmith** [comment]
> I disagree with this. As you say, the selectors are out of the scope of this PR. They were just mistakenly removed several commits ago and now they have been re-introduced exactly as they were before, but without any modifications. So I see no advantage in doing the selector changes here, compared to doing it on a separated PR right after.
> 
> For this reason, and since this is a big enough PR already, can we stick to the original plan and focus only on the tuning side to have this in asap?

Sure, I will create a new issue

> I agree with the part of not making arbitrary choices: let's have all the user-level elements at the same level.
> However, I would still make a clear distinction between tuning and selection families.
> Just like you have to do from numpy.random import randint, random instead of from numpy import randint, random, or from sklearn.ensemble import RandomForestClassifier instead of from sklearn import RandomForestClassifier.

Sure, let's use `btb.tuning` and `btb.selection` as the points to re-export functionality for end-users.

> The way to go is always using the standard LOGGER = logging.getLogger(__name__), which is the only option that gives complete control to the user to select which level each module has, as well as information of what is logging what directly at the logger name level.

I find that the advantage of `from btb.util.log import logger` is that logging can be applied consistently across the entire project. The advantage of creating a new logger in each module given the module `__name__` is if the user wants to apply logging filters at the module level. I don't anticipate that users of BTB would ever desire this, as it is an advanced usage pattern.

File: setup.cfg [hunk 1]
```diff
@@ -1,18 +1,21 @@
 [bumpversion]
-current_version = 0.3.0-dev
+current_version = 0.3.0.dev0
 commit = True
 tag = True
-parse = (?P<major>\d+)\.(?P<minor>\d+)\.(?P<patch>\d+)(\-(?P<release>[a-z]+))?
+parse = (?P<major>\d+)\.(?P<minor>\d+)\.(?P<patch>\d+)(\.(?P<release>[a-z]+)(?P<candidate>\d+))?
```
[the remark this answers is not in the document]
Comment (csala) [reply setup.cfg]: @micahjsmith and I already discussed this somewhere else and did not reach a final conclusion. Let's leave this as it is right now and, if necessary, we will change this later on.

File: btb/tuning/tuners/base.py [hunk 2]
```diff
@@ -238,4 +238,103 @@ def record(self, trials, scores):
             raise ValueError('The amount of trials must be equal to the amount of scores.')
 
         self.trials = np.append(self.trials, trials, axis=0)
-        self.scores = np.append(self.scores, scores)
+        self._trials_set.update(map(tuple, trials))
+        self.raw_scores = np.append(self.raw_scores, scores)
+        self.scores = self.raw_scores if self.maximize else -self.raw_scores
+
+
+class BaseMetaModelTuner(BaseTuner, BaseMetaModel, BaseAcquisitionFunction):
```
[the remark this answers is not in the document]
Comment (pvk-developer) [reply btb/tuning/tuners/base.py]: As far as I understand, we will always use an `acquisition` function for a `Tuner` based on a `Model`. If there is a case where we will have just a model that doesn't requiere `acquisition` then we should probably have another class that inherits from `BaseTuner` and `BaseMetaModel` only...
However if you would like to suggest a proper naming for this class I'm happy to hear. (Or just name it `BaseMetaModelAcquisitionTuner`?)

File: btb/tuning/tuners/base.py [hunk 3]
```diff
@@ -238,4 +238,103 @@ def record(self, trials, scores):
...
+        min_trials (int):
+            Number of recorded ``trials`` needed to perform a fitting over the model.
+            Defaults to 2.
+    """
+
+    def __init__(self, tunable, maximize=True, num_candidates=1000, min_trials=2, **kwargs):
+        self._num_candidates = num_candidates
+        self._min_trials = min_trials
+        super().__init__(tunable, maximize)
+        self.__init_metamodel__(**kwargs)
```
[the remark this answers is not in the document]
Comment (pvk-developer) [reply btb/tuning/tuners/base.py]: I belive that we could apply this (similar to what we did with tuners):
```
class BaseMetaModelTuner(...):
    _metamodel_kwargs = None
    _acquisition_kwargs = None

    def __init__(self, ...):
        ...
        self.__init_metamodel__(self, **(self._metamodel_kwargs or dict()))
        self.__init_acquisition__(self, **(self._acquisition_kwargs or dict()))

class MyTuner(BaseMetaModelTuner, ScaleGPTuner, ScaleAcquisition):
    def __init__(self, tunable, gp_scale, acquire_scale):
        self._metamodel_kwargs = {'scale': gp_scale}
        self._acquisition_kwargs = {'scale': acquire_scale}
        super().__init__(self, tunable)
```

File: btb/tuning/acquisition/predicted_score.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# -*- coding: utf-8 -*-
+
+from btb.tuning.acquisition.base import BaseAcquisition
+
+
+class PredictedScoreAcquisition(BaseAcquisition):
+
+    def _acquire(self, candidates, num_candidates=1):
+        candidates = candidates if len(candidates.shape) == 1 else candidates[:, 0]
```
Comment (micahjsmith) [inline btb/tuning/acquisition/predicted_score.py]: use `candidates.ndim == 1`

File: README.md [hunk 5]
```diff
@@ -162,21 +162,26 @@ we want to try out, as well as the ``Selector`` instance.
 ```python
 >>> from sklearn.ensemble import RandomForestClassifier
 >>> from sklearn.svm import SVC
+>>> from btb.selection import UCB1
```
[the remark this answers is not in the document]
Comment (micahjsmith) [reply README.md]: Will resolve in #140

File: btb/tuning/acquisition/argsort.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+from btb.tuning.acquisition.base import BaseAcquisitionFunction
+
+
+class ArgSortAcquisition(BaseAcquisitionFunction):
+    """ArgSortAcquisition class."""
+
+    def _acquire(self, candidates, num_candidates=1):
+        scores = candidates if len(candidates.shape) == 1 else candidates[:, 0]
+        sorted_scores = list(reversed(np.argsort(scores)))
```
[the remark this answers is not in the document]
Comment (pvk-developer) [reply btb/tuning/acquisition/argsort.py]: This is implemented now.
